Extract a Superblock module

emillon opened the pull request:
Hi,

Raw cstructs are a bit difficult to work with, so here's a proposition to make this a bit easier.

This wraps the superblock into its own module: there is an abstract value representing a well-sized cstruct. This can be read from a bigger cstruct (corresponding to the former `_sb_io` function), and only a small set of operations can be done on it (read params, read info, format).

What do you think about this?

In addition, I think we can go a bit further but I'm not too sure about the following:
- why are the parameters first read in `read_params`, then checked in `read`? My understanding is that the same disk is passed, so the result is always the same (and so, params do not need to be passed to `read`)
- I wanted to extract this module to an external file but I am not too sure about the external interface. The library is not wrapped, so I guess we'd need to name this `Wodan_superblock`. What about exceptions? Would it be ok to return a result type from `Wodan_superblock` and have `Wodan` translate it to the correct exceptions?

Thanks!

samoht commented:
> The library is not wrapped

It should probably be made wrapped, then :-)

emillon commented:
I agree that something should be done about that, but if it gets wrapped we'll need a way to test the internal APIs. So, either private reexports or `ppx_expect`. (we can discuss this in a different issue)

pascutto commented:
Thank you ! 

- Your first point is a good one, maybe @g2p can answer that ?
- About the separate file, I guess the library should be wrapped, as already mentioned in #50, in the meantime I think a `wodan_superblock.ml` is good, and we'll rename it later.
- We don't have a policy about exceptions for now, we should discuss that to clarify how the modules should behave, I'm fine with returning a result type when the exceptions can occur in a normal run (eg corrupted data), but I guess we don't need them for bug related ones (eg in Bitv or Keyedmap).

Also, can we wrap `_get_superblock_io` into the module ?

pascutto commented:
The API for the `Superblock` module could also be documented.

emillon commented:
Thanks for your comments.

- I moved this to `Wodan_superblock`
- I moved `_get_superblock_io` to this new module. it turns out that this can be abstracted behind a `read` and a `write` function.
- the `sub` done by `_sb_io` seems superfluous since the cstruct always has the correct size. I removed it (this is type safe).
- similarly, the "fanning" strategy does not seem to be useful, since only 512 bytes are read. I suggest to always use `fun x -> [x]`

I'll add API docs.

emillon commented:
It's also a good time to add new tests, especially now that there's a good chunk of pure code in here :)

emillon commented:
I added some tests. I haven't written API docs yet since if we end up simplifying the params handling, they'll be considerably simplified.

emillon commented:
@g2p I don't understand how the current code can raise `BadParams`. Is this check useful?

g2p commented:
read can be called without going through read_params, so checks in read are necessary.
There are two ways to use the library; either you enforce the parameters in the application, or you read them from disk and load that, read_params is only useful in the second case.

emillon commented:
Ok, thanks for the explanation!

Since wodan is wrapped now, we need an escape hatch to run the tests. We can do either:
- inline tests with `ppx_expect`
- private reexport (have an `Impl` or something module in `Wodan` that is marked as private API and is only used in tests)

What would work better for you?

g2p commented:
I think I prefer the ppx_expect option, I wouldn't want escape hatches to be available, someone might use them and they might ossify.

emillon commented:
OK! I converted the tests to `ppx_expect`, but at the moment it's not compatible because of checkseum (https://github.com/mirage/checkseum/issues/22). I added a commit to make it build by depending on the ocaml version, but this is not a viable strategy in the long run obviously.

g2p commented:
checkseum is performance-critical, so this is useful for the PR's CI but I'll wait for the checkseum fix.

g2p commented:
I'd be fine with only supporting the C variants, until Dune and Mirage have support for proper variants.

emillon commented:
I rebased and used `checkseum.c`. I don't know how to deal with the checkseum release.
